### PR TITLE
weight matrix in optimization

### PR DESCRIPTION
**Pull request by xukuanHIT**

### CODE REVIEW

File: pypose/optim/optimizer.py [hunk 1]
```diff
@@ -371,7 +378,8 @@ def __init__(self, model, solver=None, strategy=None, kernel=None, corrector=Non
         self.strategy = TrustRegion() if strategy is None else strategy
         defaults = {**{'min':min, 'max':max}, **self.strategy.defaults}
         super().__init__(model.parameters(), defaults=defaults)
-        self.jackwargs = {'vectorize': vectorize, 'flatten': True}
+        # self.jackwargs = {'vectorize': vectorize, 'flatten': True}
+        self.jackwargs = {'vectorize': vectorize, 'flatten': False}
```
Comment (wang-chen) [inline pypose/optim/optimizer.py:382]: `flatten` is `False` doesn't support multiple Model parameters, please change back to `True`. will add an additional test for this.

File: pypose/optim/optimizer.py [hunk 2]
```diff
@@ -56,14 +56,15 @@ def model_forward(self, input):
...
-        else:
-            residual = error
-        return residual.squeeze(-1)
+        # error = (output if target is None else output - target).unsqueeze(-1)
+        # if weight is not None:
+        #     residual = self.validate(weight) @ error
+        # elif hasattr(self, 'weight'):
+        #     residual = self.weight @ error
+        # else:
+        #     residual = error
```
Comment (wang-chen) [inline pypose/optim/optimizer.py:65]: remove unnecessary lines.

**Review by wang-chen — CHANGES_REQUESTED** [review]
1. remove all unnecessary lines including commented lines and `RobustModel.validate`
2. `weight` can also be given to `optimizer`, and now `Robust Model` doesn't accept `weight` anymore, you need to register 'weight' it in `optimizer`.
3. `[i for i in range(len(J.shape)-1)]` is `list(range(N-1))`
4. `J.reshape` will copy memory, do you really need that?
5. Update documentation as modified.

File: pypose/optim/optimizer.py [hunk 3]
```diff
@@ -451,17 +434,23 @@ def step(self, input, target=None, weight=None):
...
         for pg in self.param_groups:
+            weight = self.weight if weight is None else weight
             R = self.model(input, target, weight)
             J = modjac(self.model, input=(input, target, weight), **self.jackwargs)
             R, J = self.corrector(R = R, J = J)
             self.last = self.loss = self.loss if hasattr(self, 'loss') \
                                     else self.model.loss(input, target, weight)
-            A, self.reject_count = J.T @ J, 0
+            J_T = J.reshape(tuple(R.shape)+(-1,))
+            J_T = J_T.permute([len(J_T.shape)-1,]+list(range(len(J_T.shape)-1)))
```
Comment (wang-chen) [inline pypose/optim/optimizer.py:444]: why do you want the dimension of J to be rotated?
and `len(J_T.shape)` is `J_T.ndim`

**Review by wang-chen — CHANGES_REQUESTED** [review]

File: pypose/optim/optimizer.py (hunk 3 above)
Comment (xukuanHIT) [reply pypose/optim/optimizer.py:444]: > len(

Screenshot from 2022-12-17 16-59-34

File: pypose/optim/optimizer.py [hunk 4]
```diff
@@ -451,17 +434,28 @@ def step(self, input, target=None, weight=None):
...
         for pg in self.param_groups:
+            weight = self.weight if weight is None else weight
             R = self.model(input, target, weight)
             J = modjac(self.model, input=(input, target, weight), **self.jackwargs)
             R, J = self.corrector(R = R, J = J)
             self.last = self.loss = self.loss if hasattr(self, 'loss') \
                                     else self.model.loss(input, target, weight)
-            A, self.reject_count = J.T @ J, 0
+            # Let Nr: number of residuals, Dr: dimension of each residual, Np: number of model parameters
+            # Now, the shape of R: [Nr, Dr], J: [Nr*Dr, Np], weight: [Nr, Dr, Dr]
```
Comment (wang-chen) [inline pypose/optim/optimizer.py:444]: This is not correct, `R` can be any higher dimension, depending on the model defined by the users. I will and a test example to the test folder.

**Comment by xukuanHIT** [comment]
The reason why the loss didn't decrease in the previous tests is that we can get the optimal solution in one step. Now I have changed "test_optim_anybatch" to "multi-parameter", so we can see the explicit loss decrease.

File: pypose/optim/optimizer.py [hunk 5]
```diff
@@ -451,22 +434,28 @@ def step(self, input, target=None, weight=None):
...
+            weight = self.weight if weight is None else weight
+            R = self.model(input, target)
+            J = modjac(self.model, input=(input, target), **self.jackwargs)
             R, J = self.corrector(R = R, J = J)
             self.last = self.loss = self.loss if hasattr(self, 'loss') \
-                                    else self.model.loss(input, target, weight)
-            A, self.reject_count = J.T @ J, 0
+                                    else self.model.loss(input, target)
+            J_T = J.reshape(tuple(R.shape)+(-1,))
+            J_T = J_T.permute([J_T.ndim-1,]+list(range(J_T.ndim-1)))
```
Comment (wang-chen) [inline pypose/optim/optimizer.py:444]: How about changing those two lines to:
```python
J_T = J.T.reshape((-1,) + R.shape)
```

**Review by wang-chen — APPROVED** [review]
